src/hotspot/cpu/s390/c2_MacroAssembler_s390.cpp

```diff
@@ -46,20 +46,41 @@
-// by a few bytes. The result always indicates the number of copied characters.
-// When used as a character index, the returned value points to the first incompatible character.
 //
-// Note: Does not behave exactly like package private StringUTF16 compress java implementation in case of failure:
-// - Different number of characters may have been written to dead array (if precise is false).
-// - Returns a number <cnt instead of 0. (Result gets compared with cnt.)
+// The result indicates success or failure of the operation.
+//   General compress operation (cut off high order byte which must be all zeroes).
+//    = len - all characters have been successfully compressed.
+//    = 0   - compress failed. At least one character was found with a non-zero high order byte.
+//            This is the failure return value which exactly corresponds to the Java implementation.
+//    0 <= result < len - compress failed. That many characters were compressed successfully
+//                        before the first non-compressable character was found. This is the
+//                        current, but not fully compatible, implementation. See below.
+//   Encode to ISO or 7-bit ASCII array.
+//    = len - all characters have been encoded successfully.
+//    < len - encode failed. That many characters were encoded successfully.
+//            When used as an index into the character array, the return value addresses the
+//            first not encodeable character.
+//
+// If precise is true, the processing stops exactly at the point where a failure is detected.
+// More characters than indicated by the return value may have been read from the src array.
+// Exactly the number of characters indicated by the return value have been written to dst.
+// If precise is false, a few characters more than indicated by the return value may have been
+// written to the dst array. In any failure case, The result value indexes the first invalid character.
 unsigned int C2_MacroAssembler::string_compress(Register result, Register src, Register dst, Register cnt,
-                                                Register tmp,    bool precise) {
+                                                Register tmp,    bool precise, bool toASCII) {
   assert_different_registers(Z_R0, Z_R1, result, src, dst, cnt, tmp);
 
+  unsigned short char_mask = 0xff00;  // all selected bits must be '0' for a char to be valid
+  unsigned int   mask_ix_l = 0;       // leftmost one bit pos in mask
+  unsigned int   mask_ix_r = 7;       // rightmost one bit pos in mask
   if (precise) {
-    BLOCK_COMMENT("encode_iso_array {");
+    if (toASCII) {
+      BLOCK_COMMENT("encode_ascii_array {");
+      char_mask = 0xff80;
+      mask_ix_r = 8;         // rightmost one bit pos in mask. ASCII only uses codes 0..127
+    } else {
+      BLOCK_COMMENT("encode_iso_array {");
+    }
   } else {
     BLOCK_COMMENT("string_compress {");
```

Comment:
Maybe assert(!toASCII, "unsupported") ?

Reply:
Hmm,
could be done. On the other hand, calls are exclusively from ad-file match rules. But ok,  I'll create another commit.